drivers/wdm/audio/backpln/portcls/irpstream2.cpp

```diff
@@ -0,0 +1,597 @@
+            HeaderLength = Header->DataUsed;
+        }
+        
+        // append page count
+        StreamPageCount += ADDRESS_AND_SIZE_TO_SPAN_PAGES(
+            MmGetMdlByteOffset(Mdl), HeaderLength);
+
+        // move to next header / mdl
+        Mdl = Mdl->Next;
+        Header = (PKSSTREAM_HEADER)((ULONG_PTR)Header + Header->Size);
+
+    }while(Length);
+
+    // allocate stream data
+    StreamData = (PKSSTREAM_DATA)AllocateItem(NonPagedPool, sizeof(KSSTREAM_DATA) + 
+        sizeof(KSSTREAM_TAG) * StreamPageCount, TAG_PORTCLASS);
+    if (!StreamData)
+    {
+        // done
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    // now get a system address for the user buffers
+    Header = (PKSSTREAM_HEADER)Irp->AssociatedIrp.SystemBuffer;
+    Length = IoStack->Parameters.DeviceIoControl.OutputBufferLength;
+    Mdl = Irp->MdlAddress;
+    
+    do
+    {
+        /* subtract size */
+        Length -= Header->Size;    
+
+        /* get system address */
+        char* Data = (char*)MmGetSystemAddressForMdlSafe(Mdl, NormalPagePriority);
+
+        /* check for success */
+        if (!Data)
+        {
+            FreeItem(StreamData, TAG_PORTCLASS);
+            // done
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+        
+        if (m_Descriptor->DataFlow == KSPIN_DATAFLOW_IN)
+        {
+            HeaderLength = Header->DataUsed;
+        }
+        else if (m_Descriptor->DataFlow == KSPIN_DATAFLOW_OUT)
+        {
+            HeaderLength = Header->DataUsed;
+        }
+        
+        // append pages
+        while(HeaderLength) 
+        {
+            char* NextData = (char*)ROUND_TO_PAGES(Data+1);
+            LONG PageSize = NextData - Data;
+            
+            if(PageSize > HeaderLength) 
+                PageSize = HeaderLength;
+            HeaderLength -= PageSize;
+            
+            StreamData->Tags[StreamData->nTags].Data = Data;
+            StreamData->Tags[StreamData->nTags].Length = PageSize;
+            StreamData->nTags++;
+            
+            Data = NextData;
+        }
+
+        // move to next header / mdl
+        Mdl = Mdl->Next;
+        Header = (PKSSTREAM_HEADER)((ULONG_PTR)Header + Header->Size);
+        
+    }while(Length);
+
+    // store stream data
+    Irp->Tail.Overlay.DriverContext[STREAM_DATA_OFFSET] = (PVOID)StreamData;
+
+    *Data = TotalStreamData;
+
+    // mark irp as pending
+    IoMarkIrpPending(Irp);
+
+    // add irp to cancelable queue
+    KsAddIrpToCancelableQueue(&m_IrpList, &m_IrpListLock, Irp, KsListEntryTail, NULL);
+
+    // disable mapping failed status
+    m_OutOfMapping = FALSE;
+
+    // done
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+NTAPI
+CIrpQueue2::GetMapping(
+    OUT PUCHAR * Buffer,
+    OUT PULONG BufferSize)
+{
+    ASSERT(0);
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+VOID
+NTAPI
+CIrpQueue2::UpdateMapping(
+    IN ULONG BytesWritten)
+{
+    ASSERT(0);
```

Comment:
```suggestion
    UNIMPLEMENTED;
```

Reply:
assert(0) is fine to see if it's being used while working on something.

Reply:
OK.